src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -765,7 +765,146 @@ private ITypeReference GetUnionTypeType(UnionTypeSyntax syntax)
+                            .DiscriminatorPropertyMemberDuplicatedValue(discriminatorPropertyName, discriminatorMemberValue));
+
+                        continue;
+                    }
+
+                    memberDiscriminatorValues.Add(discriminatorMemberValue);
+                    expandedMemberTypes.Add(objectType);
+                }
+                else if (memberTypeEvaluated is DiscriminatedObjectType memberDiscriminatedObjectType)
+                {
+                    // validate it has the same discriminator property
+                    if (!string.Equals(discriminatorPropertyName, memberDiscriminatedObjectType.DiscriminatorProperty.Name, LanguageConstants.IdentifierComparison))
+                    {
+                        errorDiagnostics.Add(
+                            DiagnosticBuilder.ForPosition(memberSyntax)
+                                .DiscriminatorPropertyNameMustMatch(discriminatorPropertyName));
+
+                        continue;
+                    }
+
+                    // validate there's not value overlap
+                    var nestedHasError = false;
+
+                    foreach (var nestedDiscriminatorValue in memberDiscriminatedObjectType.UnionMembersByKey.Keys)
+                    {
+                        if (memberDiscriminatorValues.Contains(nestedDiscriminatorValue))
+                        {
+                            errorDiagnostics.Add(
+                                DiagnosticBuilder.ForPosition(memberSyntax)
+                                .DiscriminatorPropertyMemberDuplicatedValue(discriminatorPropertyName, nestedDiscriminatorValue));
+
+                            nestedHasError = true;
+                        }
+
+                        memberDiscriminatorValues.Add(nestedDiscriminatorValue);
+                    }
+
+                    if (!nestedHasError)
+                    {
+                        expandedMemberTypes.AddRange(memberDiscriminatedObjectType.UnionMembersByKey.Values);
+                    }
+                }
+                else
+                {
+                    return ErrorType.Create(
+                        DiagnosticBuilder.ForPosition(discriminatorDecorator)
+                            .DiscriminatorDecoratorOnlySupportedForObjectUnions());
+                }
+            }
+
+            if (errorDiagnostics.Any())
+            {
+                return ErrorType.Create(errorDiagnostics);
+            }
+
+            var discriminatedObjectType = new DiscriminatedObjectType(
+                name: string.Join(" | ", TypeHelper.GetOrderedTypeNames(expandedMemberTypes)),
+                validationFlags: TypeSymbolValidationFlags.Default,
+                discriminatorKey: discriminatorPropertyName,
+                unionMembers: expandedMemberTypes);
+
+            return IsNullableUnion(unionTypeSyntax)
+                ? TypeHelper.CreateTypeUnion(discriminatedObjectType, LanguageConstants.Null)
```

Comment:
This case is already be handled by `ConvertTypeExpressionToType(NullableTypeSyntax)`. For any "wrapper" type syntax (like `NullableTypeSyntax`, `ArrayTypeSyntax`, `NonNullAssertionSyntax`, `ParenthesizedExpressionSyntax`, etc.), the wrapper syntax should get the type of the inner expression and then modify it (rather than having the inner expression try to figure out what wrapper types should be applied to it).